In [2]:
import pandas as pd
import numpy as np

Prepare a Naive Bayes (NB) classification model using the dataset provided. Your task includes the following steps:

1. Load and explore the dataset.

In [3]:
df = pd.read_csv('fake_or_real_news.csv')
df

C:\Users\G M RIFAT REZA\AppData\Local\Temp\ipykernel_17688\3901040168.py:1: DtypeWarning: Columns (22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('fake_or_real_news.csv')


text label Unnamed: 2  \
0     Daniel Greenfield, a Shillman Journalism Fello...  FAKE        NaN   
1     Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE        NaN   
2     U.S. Secretary of State John F. Kerry said Mon...  REAL        NaN   
3     — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE        NaN   
4     It's primary day in New York and front-runners...  REAL        NaN   
...                                                 ...   ...        ...   
7790  The State Department told the Republican Natio...  REAL        NaN   
7791  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...  FAKE        NaN   
7792   Anti-Trump Protesters Are Tools of the Oligar...  FAKE        NaN   
7793  ADDIS ABABA, Ethiopia —President Obama convene...  REAL        NaN   
7794  Jeb Bush Is Suddenly Attacking Trump. Here's W...  REAL        NaN   

     Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6 Unnamed: 7 Unnamed: 8  \
0           NaN        NaN        NaN        NaN        NaN        NaN   
1           NaN        NaN        NaN        NaN        NaN        NaN   
2           NaN        NaN        NaN        NaN        NaN        NaN   
3           NaN        NaN        NaN        NaN        NaN        NaN   
4           NaN        NaN        NaN        NaN        NaN        NaN   
...         ...        ...        ...        ...        ...        ...   
7790        NaN        NaN        NaN        NaN        NaN        NaN   
7791        NaN        NaN        NaN        NaN        NaN        NaN   
7792        NaN        NaN        NaN        NaN        NaN        NaN   
7793        NaN        NaN        NaN        NaN        NaN        NaN   
7794        NaN        NaN        NaN        NaN        NaN        NaN   

     Unnamed: 9  ... Unnamed: 129 Unnamed: 130 Unnamed: 131 Unnamed: 132  \
0           NaN  ...          NaN          NaN          NaN          NaN   
1           NaN  ...          NaN          NaN          NaN          NaN   
2           NaN  ...          NaN          NaN          NaN          NaN   
3           NaN  ...          NaN          NaN          NaN          NaN   
4           NaN  ...          NaN          NaN          NaN          NaN   
...         ...  ...          ...          ...          ...          ...   
7790        NaN  ...          NaN          NaN          NaN          NaN   
7791        NaN  ...          NaN          NaN          NaN          NaN   
7792        NaN  ...          NaN          NaN          NaN          NaN   
7793        NaN  ...          NaN          NaN          NaN          NaN   
7794        NaN  ...          NaN          NaN          NaN          NaN   

     Unnamed: 133 Unnamed: 134 Unnamed: 135 Unnamed: 136 Unnamed: 137  \
0             NaN          NaN          NaN          NaN          NaN   
1             NaN          NaN          NaN          NaN          NaN   
2             NaN          NaN          NaN          NaN          NaN   
3             NaN          NaN          NaN          NaN          NaN   
4             NaN          NaN          NaN          NaN          NaN   
...           ...          ...          ...          ...          ...   
7790          NaN          NaN          NaN          NaN          NaN   
7791          NaN          NaN          NaN          NaN          NaN   
7792          NaN          NaN          NaN          NaN          NaN   
7793          NaN          NaN          NaN          NaN          NaN   
7794          NaN          NaN          NaN          NaN          NaN   

     Unnamed: 138  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN  
...           ...  
7790          NaN  
7791          NaN  
7792          NaN  
7793          NaN  
7794          NaN  

[7795 rows x 139 columns]

2. Preprocess the data if required (handle missing values, encode categorical variables, etc.).

In [4]:
df.isnull().sum()

text             866
label           1040
Unnamed: 2      7477
Unnamed: 3      7554
Unnamed: 4      7616
                ... 
Unnamed: 134    7794
Unnamed: 135    7794
Unnamed: 136    7794
Unnamed: 137    7794
Unnamed: 138    7794
Length: 139, dtype: int64

In [5]:
df_cleaned1 = df.loc[:, ~df.columns.str.startswith('Unnamed:')]
df_cleaned1

text label
0     Daniel Greenfield, a Shillman Journalism Fello...  FAKE
1     Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE
2     U.S. Secretary of State John F. Kerry said Mon...  REAL
3     — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE
4     It's primary day in New York and front-runners...  REAL
...                                                 ...   ...
7790  The State Department told the Republican Natio...  REAL
7791  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...  FAKE
7792   Anti-Trump Protesters Are Tools of the Oligar...  FAKE
7793  ADDIS ABABA, Ethiopia —President Obama convene...  REAL
7794  Jeb Bush Is Suddenly Attacking Trump. Here's W...  REAL

[7795 rows x 2 columns]

In [6]:
df_cleaned1.isnull().sum()

text      866
label    1040
dtype: int64

In [7]:
df_cleaned2 = df_cleaned1.dropna().reset_index(drop=True)
df_cleaned2

text label
0     Daniel Greenfield, a Shillman Journalism Fello...  FAKE
1     Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE
2     U.S. Secretary of State John F. Kerry said Mon...  REAL
3     — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE
4     It's primary day in New York and front-runners...  REAL
...                                                 ...   ...
6749  The State Department told the Republican Natio...  REAL
6750  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...  FAKE
6751   Anti-Trump Protesters Are Tools of the Oligar...  FAKE
6752  ADDIS ABABA, Ethiopia —President Obama convene...  REAL
6753  Jeb Bush Is Suddenly Attacking Trump. Here's W...  REAL

[6754 rows x 2 columns]

In [8]:
df_cleaned2.isnull().sum()

text     0
label    0
dtype: int64

In [9]:
df_final = df_cleaned2[df_cleaned2['label'].isin(['REAL', 'FAKE'])].reset_index(drop=True)
df_final

text label
0     Daniel Greenfield, a Shillman Journalism Fello...  FAKE
1     Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE
2     U.S. Secretary of State John F. Kerry said Mon...  REAL
3     — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE
4     It's primary day in New York and front-runners...  REAL
...                                                 ...   ...
6310  The State Department told the Republican Natio...  REAL
6311  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...  FAKE
6312   Anti-Trump Protesters Are Tools of the Oligar...  FAKE
6313  ADDIS ABABA, Ethiopia —President Obama convene...  REAL
6314  Jeb Bush Is Suddenly Attacking Trump. Here's W...  REAL

[6315 rows x 2 columns]

In [10]:
df_final.shape

(6315, 2)

3. Split the data into training and testing sets.

In [11]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(df_final.text, df_final.label, test_size=0.2, random_state=42)

In [12]:
X_train

2304    At 10:16 p.m. on Aug. 15, 1977, astronomer Jer...
5517    Kaboom! Meteor Turns Night Into Day in Arizona...
1086    We Are Change \nIn this video Luke Rudkowski b...
3316    October 28, 2016 Let’s Be Clear—A Vote for War...
1172    Sanders was on ‘Meet the Press’ again Sunday t...
                              ...                        
3772    Republican presidential candidate Donald Trump...
5191    Notify me of follow-up comments by email. Noti...
5226    Yemeni President Abed Rabbo Mansour Hadi has r...
5390    Let’s Be Clear – A Vote For Warmonger Hillary ...
860     David M. Perry is an associate professor of hi...
Name: text, Length: 5052, dtype: object

In [13]:
y_train

2304    REAL
5517    FAKE
1086    FAKE
3316    FAKE
1172    REAL
        ... 
3772    REAL
5191    FAKE
5226    REAL
5390    FAKE
860     REAL
Name: label, Length: 5052, dtype: object

4. Build a Naive Bayes model using the training data.

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
v = CountVectorizer()

In [15]:
X_train_cv = v.fit_transform(X_train.values)
X_train_cv

<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 1684498 stored elements and shape (5052, 60181)>

In [16]:
X_train_cv.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 1, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [17]:
v.vocabulary_

{'at': 5103,
 '10': 116,
 '16': 357,
 'on': 37988,
 'aug': 5307,
 '15': 315,
 '1977': 612,
 'astronomer': 5079,
 'jerry': 28929,
 'ehman': 17777,
 'culled': 13696,
 'through': 53568,
 'data': 14185,
 'collected': 11526,
 'by': 8998,
 'the': 53268,
 'super': 51780,
 'powered': 41382,
 'big': 7047,
 'ear': 17456,
 'radio': 43155,
 'telescope': 52949,
 'in': 26926,
 'delaware': 14761,
 'ohio': 37856,
 'and': 3809,
 'there': 53362,
 'it': 28485,
 'was': 57831,
 'staring': 50674,
 'back': 5685,
 'something': 49798,
 'seti': 47931,
 'search': 47496,
 'for': 21342,
 'extraterrestrial': 19712,
 'intelligence': 27823,
 'investigators': 28188,
 'had': 24113,
 'never': 36653,
 'found': 21558,
 'before': 6524,
 'or': 38178,
 'since': 48889,
 'transcript': 54340,
 'revealed': 45270,
 'an': 3741,
 'astonishingly': 5059,
 'strong': 51311,
 'non': 37038,
 'terrestrial': 53112,
 'signal': 48755,
 'which': 58285,
 'didn': 15644,
 'even': 19170,
 'originate': 38291,
 'our': 38434,
 'solar': 49715,
 'syst

In [18]:
from sklearn.naive_bayes import MultinomialNB

model = MultinomialNB()
model.fit(X_train_cv, y_train)

MultinomialNB()

5. Evaluate the model's performance using appropriate metrics (e.g., accuracy, confusion matrix).

In [19]:
X_test_cv = v.transform(X_test)

In [20]:
model.score(X_test_cv, y_test)

0.894695170229612

In [21]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

X_test_cv = v.transform(X_test)

y_pred = model.predict(X_test_cv)

accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:\n", cm)

report = classification_report(y_test, y_pred)
print("Classification Report:\n", report)


Accuracy: 0.894695170229612
Confusion Matrix:
 [[530  86]
 [ 47 600]]
Classification Report:
               precision    recall  f1-score   support

        FAKE       0.92      0.86      0.89       616
        REAL       0.87      0.93      0.90       647

    accuracy                           0.89      1263
   macro avg       0.90      0.89      0.89      1263
weighted avg       0.90      0.89      0.89      1263



6. Interpret the results and provide a brief summary of findings.

Model Performance Summary

**Accuracy:**

The model achieved an accuracy of ~89.5% on the test set, meaning it correctly classified about 9 out of 10 news articles.

**Confusion Matrix Analysis:**

[[530  86]
 [ 47 600]]


**FAKE news:**

530 correctly identified as FAKE (True Positives)

86 misclassified as REAL (False Negatives)

**REAL news:**

600 correctly identified as REAL (True Positives)

47 misclassified as FAKE (False Negatives)

**Precision, Recall, F1-Score:**

Class	Precision	Recall	F1-Score
FAKE	0.92	0.86	0.89
REAL	0.87	0.93	0.90

**FAKE class:**

High precision (0.92) → Most predicted FAKE articles are truly FAKE.

Slightly lower recall (0.86) → Some FAKE articles are misclassified as REAL.

**REAL class:**

High recall (0.93) → Most REAL articles are correctly identified.

Slightly lower precision (0.87) → Some REAL articles are misclassified as FAKE.

**Overall Interpretation:**

The model is fairly balanced between FAKE and REAL classes.

Misclassifications exist but are relatively low, showing Naive Bayes is effective for text classification in detecting fake news.

The bag-of-words approach captures key words and patterns distinguishing FAKE from REAL articles.

**✅ Conclusion:**

* The Multinomial Naive Bayes classifier performs well for fake news detection using textual features.

* It can be confidently used for preliminary classification of news articles with ~89–90% accuracy.

* Some improvements may be possible using TF-IDF features, stop-word removal, or more advanced models like Logistic Regression or Transformers for higher accuracy.

In [ ]:
df.isnull()

(7795, 139)

In [ ]:
df.shape

In [ ]:
print('The value is')
a = 1 